### ISSUE TITLE
False Positive: needless_raw_string_hashes

### ISSUE THREAD

**Issue by FintanH** [rust-lang/rust-clippy#11402]
### Summary

On our `radicle-surf` project, the lint job failed due to this clippy warning. This is the [job](https://github.com/radicle-dev/radicle-git/actions/runs/5965259965/job/16182278180?pr=139) and the offending code is:

```rust
#[test]
fn test_both_missing_eof_newline() {
    let buf = r#"
diff --git a/.env b/.env
index f89e4c0..7c56eb7 100644
--- a/.env
+++ b/.env
@@ -1 +1 @@
-hello=123
\ No newline at end of file
+hello=1234
\ No newline at end of file
"#;
    let diff = git2::Diff::from_buffer(buf.as_bytes()).unwrap();
    let diff = Diff::try_from(diff).unwrap();
    assert_eq!(
        diff.modified().next().unwrap().diff.eof(),
        Some(EofNewLine::BothMissing)
    );
}
```

What's strange is there is a very similar test below that does not set off the clippy warning:
```rust
#[test]
fn test_none_missing_eof_newline() {
    let buf = r#"
diff --git a/.env b/.env
index f89e4c0..7c56eb7 100644
--- a/.env
+++ b/.env
@@ -1 +1 @@
-hello=123
+hello=1234
"#;
    let diff = git2::Diff::from_buffer(buf.as_bytes()).unwrap();
    let diff = Diff::try_from(diff).unwrap();
    assert_eq!(
        diff.modified().next().unwrap().diff.eof(),
        Some(EofNewLine::NoneMissing)
    );
}
```

### Lint Name

needless_raw_string_hashes

### Reproducer

I tried this code:

```rust
#[test]
fn test_both_missing_eof_newline() {
    let buf = r#"
diff --git a/.env b/.env
index f89e4c0..7c56eb7 100644
--- a/.env
+++ b/.env
@@ -1 +1 @@
-hello=123
\ No newline at end of file
+hello=1234
\ No newline at end of file
"#;
    let diff = git2::Diff::from_buffer(buf.as_bytes()).unwrap();
    let diff = Diff::try_from(diff).unwrap();
    assert_eq!(
        diff.modified().next().unwrap().diff.eof(),
        Some(EofNewLine::BothMissing)
    );
}
```

I saw this happen:

```
error: unnecessary hashes around raw string literal
   --> radicle-surf/t/src/diff.rs:363:15
    |
363 |       let buf = r#"
    |  _______________^
364 | | diff --git a/.env b/.env
365 | | index f89e4c0..7c56eb7 100644
366 | | --- a/.env
...   |
372 | | \ No newline at end of file
373 | | "#;
    | |__^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_raw_string_hashes
    = note: `-D clippy::needless-raw-string-hashes` implied by `-D warnings`
help: try
    |
363 ~     let buf = r"
364 + diff --git a/.env b/.env
365 + index f89e4c0..7c56eb7 100644
366 + --- a/.env
367 + +++ b/.env
368 + @@ -1 +1 @@
369 + -hello=123
370 + \ No newline at end of file
371 + +hello=1234
372 + \ No newline at end of file
373 ~ ";
    |

error: could not compile `radicle-surf-test` (lib test) due to previous error
```

I expected for the lint job to pass. It can be reproduced in the git repository by running `scripts/ci/lint`.

### Version

```text
rustc 1.72.0 (5680fa18f 2023-08-23)
binary: rustc
commit-hash: 5680fa18feaa87f3ff04063800aec256c3d4b4be
commit-date: 2023-08-23
host: x86_64-unknown-linux-gnu
release: 1.72.0
LLVM version: 16.0.5
```

**Comment by cemoktra** [comment]
Does you fix also allow for raw strings like: `r#"escaped single quote \' is false positive"#`

**Comment by Centri3** [comment]
I fail to see how either of these are false positives. You can remove the hashes just fine.

`needless_raw_string_hashes` concerns the number of hashes, not the raw string itself.

The test that doesn't emit anything looks like a FN.

**Comment by cemoktra** [comment]
> I fail to see how either of these are false positives. You can remove the hashes just fine.
> 
> `needless_raw_string_hashes` concerns the number of hashes, not the raw string itself.
> 
> The test that doesn't emit anything looks like a FN.

No. Removing the hashes caused unit tests to start failing. So something is changed in the string

**Comment by y21** [comment]
I'm also not entirely sure what the problem here is. That suggestion looks good to me, and the snippet it suggests also compares equal to your original string, so I don't know how that caused tests to fail.
```rs
fn main() {
    let a = r#"escaped single quote \' is false positive"#;
    let b = r"escaped single quote \' is false positive"; // clippy's suggestion
    dbg!(a == b);
}
```
```
[src/main.rs:4] a == b = true
```

**Comment by Centri3** [comment]
Removing the hashes should only change the string if there is a `"` with the amount of suggested hashes following it (or more), but would also be a compile error rather than a test failure. So can you give an example of a test that fails? Because the demonstrative examples work perfectly fine for me.

**Comment by cemoktra** [comment]
We are parsing the string in this test, more i can't say. It seems to make a difference. and i just checked, the string also contains double quotes

**Comment by FintanH** [comment]
> I fail to see how either of these are false positives. You can remove the hashes just fine.
> 
> `needless_raw_string_hashes` concerns the number of hashes, not the raw string itself.
> 
> The test that doesn't emit anything looks like a FN.

I see, I didn't realise that the `#`'s weren't needed in my examples above. I guess I'm reporting a false negative then :)

**Comment by torfsen** [comment]
@FintanH: I've tried reproducing this, but so far couldn't. For example, [this reduced playground snippet](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=431694752ab859fd0f4ce773942487ff) with your two test cases produces the `needless_raw_string_hashes` for both, as expected.

Do you have a minimal, standalone example that shows the issue?

**Comment by FintanH** [comment]
> @FintanH: I've tried reproducing this, but so far couldn't. For example, [this reduced playground snippet](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=431694752ab859fd0f4ce773942487ff) with your two test cases produces the `needless_raw_string_hashes` for both, as expected.
> 
> Do you have a minimal, standalone example that shows the issue?

Well I did report it with 1.72 whereas that playground is running 1.75, so maybe that's playing a factor? :)

Judging from the conversation, it seems like the fix was to remove the `#` anyway, but I guess it wasn't clear why some cases passed and others didn't. I'm happy to consider this as resolved if it can't be reproduced.
